### PR TITLE
Feature/37186 sign on

### PR DESCRIPTION
**Pull request by GuyHarwood**
### DfE Signin integration

- [x] refactor existing auth strategies into uniform structure
- [x] add dfe sign-in strategy, as per example
- [x] local working poc 
- [x] add issuer timeout to config
- [x] link in school org id
- [x] persist new users
- [x] issuer discovery is fault tolerant
- [x] sign out implemented

## KNOWN ISSUE
Signout results in a 404 for the user when they land back at dfe signin.
This will be covered in a new ticket once we are unblocked on this

### CODE REVIEW

File: admin/authentication/dfe-signon-strategy.js [hunk 1]
```diff
@@ -0,0 +1,42 @@
+'use strict'
+
+const config = require('../config')
+const { Strategy, Issuer } = require('openid-client')
+const logger = require('../services/log.service').getLogger()
+const asyncRetry = require('login.dfe.async-retry')
+
+const initSignOn = async () => {
+  Issuer.defaultHttpOptions = { timeout: 10000 }
```
Comment (jon-shipley) [inline admin/authentication/dfe-signon-strategy.js]: Are the units milliseconds?

File: admin/authentication/dfe-signon-strategy.js [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+  })
+  if (config.Auth.dfeSignIn.clockTolerance && config.Auth.dfeSignIn.clockTolerance > 0) {
+    client.CLOCK_TOLERANCE = config.Auth.dfeSignIn.clockTolerance
+  }
+
+  return new Strategy({
+    client,
+    params: {
+      // redirect_uri: `http://localhost:3001/auth`,
+      scope: 'openid profile email'
```
Comment (jon-shipley) [inline admin/authentication/dfe-signon-strategy.js]: magic string?

File: admin/config.js [hunk 3]
```diff
@@ -136,5 +135,18 @@ module.exports = {
     Port: process.env.REDIS_PORT || 6379,
     Key: process.env.REDIS_KEY,
     useTLS: getEnvironment() !== 'Local-Dev'
+  },
+  Auth: {
+    mode: process.env.AUTH_MODE || 'Local', // see ./lib/consts/auth-modes.js for valid options
```
Comment (jon-shipley) [inline admin/config.js]: ✅

File: admin/config.js [hunk 4]
```diff
@@ -136,5 +135,18 @@ module.exports = {
...
+    dfeSignIn: {
+      authUrl: process.env.DFE_SIGNON_AUTH_URL,
+      clientId: process.env.DFE_SIGNON_CLIENT_ID,
+      clientSecret: process.env.DFE_SIGNON_CLIENT_SECRET,
+      apiSecret: process.env.DFE_SIGNON_API_SECRET,
+      clockTolerance: process.env.DFE_SIGNON_CLOCK_TOLERANCE || 300
```
Comment (jon-shipley) [inline admin/config.js]: Are the units milliseconds?

File: admin/controllers/ping.js [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+async function getPing (req, res) {
+  // get build number from /build.txt
+  // get git commit from /commit.txt
+  let buildNumber = 'NOT FOUND'
+  let commitId = 'NOT FOUND'
+  try {
+    buildNumber = await getBuildNumber()
+  } catch (error) {
+
+  }
```
Comment (jon-shipley) [inline admin/controllers/ping.js]: Your code is fine, but I quite like:

```suggestion
  } catch (ignore) {}
```

File: admin/controllers/ping.js [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+  res.setHeader('Content-Type', 'application/json')
+  let obj = {
+    'Build': buildNumber,
+    'Commit': commitId,
+    'CurrentServerTime': Date.now()
+  }
+  return res.status(200).send(obj)
+}
+
+module.exports = getPing
```
Comment (jon-shipley) [inline admin/controllers/ping.js]: ✅

File: admin/routes/index.js [hunk 7]
```diff
@@ -52,45 +52,19 @@ router.get('/privacy', (req, res) => getPrivacyPage(req, res))
...
 router.get('/ping', (req, res) => getPing(req, res))
 
-/* NCA Tools Authentication Endpoint */
+/* NCA Tools or DfeSignIn Authentication Endpoint */
 router.post('/auth',
   function (req, res, next) {
-    // Only allow post requests if NCA TOOLS is enabled
-    if (!config.NCA_TOOLS_AUTH_URL) {
+    // Only allow post requests if not in local auth mode
+    if (config.Auth.mode === authModes.local) {
```
Comment (jon-shipley) [inline admin/routes/index.js]: The comment and the code disagree here.

File: admin/authentication/dfe-signon-strategy.js (hunk 2 above)
Comment (GuyHarwood) [reply admin/authentication/dfe-signon-strategy.js]: taken verbatim from provided example.  will figure out what this is supposed to be and action appropriately

File: admin/authentication/dfe-signon-strategy.js (hunk 1 above)
Comment (GuyHarwood) [reply admin/authentication/dfe-signon-strategy.js]: taken verbatim from provided example, will formalise this soon.

File: admin/authentication/dfe-signon-strategy.js (hunk 2 above)
Comment (GuyHarwood) [reply admin/authentication/dfe-signon-strategy.js]: @jon-shipley these are the 'sets' of data you want included in the payload returned from the auth provider.  such as their email, org info etc.

This has been added as config with a default.

File: admin/authentication/dfe-signon-strategy.js (hunk 1 above)
Comment (GuyHarwood) [reply admin/authentication/dfe-signon-strategy.js]: now a formal setting with unit postfix

File: admin/config.js (hunk 4 above)
Comment (GuyHarwood) [reply admin/config.js]: seconds.  config updated with postfix

File: admin/app.js [hunk 8]
```diff
@@ -210,19 +211,32 @@ passport.deserializeUser(function (user, done) {
...
-  new LocalStrategy({
-    passReqToCallback: true
-  },
-  require('./authentication/local-strategy')
-  )
-)
+// initialise chosen auth strategy only
+switch (config.Auth.mode) {
+  case authModes.dfeSignIn:
+    ;(async function () {
```
Comment (jon-shipley) [inline admin/app.js]: This IIFE is not synchronised with the switch statement at all.  Is there any way to remove the await on the `dfe-signin-strategy` and make this synchronous?

File: admin/controllers/authentication.js [hunk 9]
```diff
@@ -27,65 +54,107 @@ const getSignIn = (req, res) => {
...
+  logger.info(`postSignIn: User ID logged in:
+    id:${req.user.id} \n
+    displayName:${req.user.displayName} \n
+    role:${req.user.role} \n
+    timezone:"${req.user.timezone}"`)
+
+  switch (req.user.role) {
+    case 'TEACHER':
+    case 'HEADTEACHER':
     case 'HELPDESK':
```
Comment (jon-shipley) [inline admin/controllers/authentication.js]: Don't the helpdesk have their own page now?

File: admin/app.js (hunk 8 above)
Comment (GuyHarwood) [reply admin/app.js]: yes, we could make the initialisation of OpenIdConnect synchronous, which is a fairly valid ask given that the app should not even start if DfE signin service is not available when enabled.

File: admin/controllers/authentication.js (hunk 9 above)
Comment (GuyHarwood) [reply admin/controllers/authentication.js]: indeed they do.  I tested the helpdesk role yesterday and it worked, however I will walk through the code today to ensure its valid.  Maybe it does a redirect after establishing that no school impersonation is set...

File: admin/services/dfe-signin.service.js [hunk 10]
```diff
@@ -6,23 +6,75 @@ const bluebird = require('bluebird')
...
+  getRoleInfo: async (user) => {
+    try {
+      // get role info...
+      const token = await createJwtForDfeApi()
+      const userInfo = await getUserInfoFromDfeApi(token, user)
+      logger.debug('response received from dfe.API:')
+      logger.debugObject(userInfo)
+      // TODO array check
+      const mtcRole = roleService.mapDfeRoleToMtcRole(userInfo.roles[0].code)
+      user.role = mtcRole
```
Comment (jon-shipley) [inline admin/services/dfe-signin.service.js]: This function takes an object as the param, so it is passed by reference.  I find it slightly misleading that it also returns the object which is then re-assigned by to `req.user` - which is the original object that was passed in.

File: admin/controllers/authentication.js [hunk 11]
```diff
@@ -59,11 +57,11 @@ const getSignIn = (req, res) => {
 }
 
 const postDfeSignIn = async (req, res) => {
-  req.user = await dfeSignInService.process(req.user)
+  req.user = await dfeSignInService.getRoleInfo(req.user)
```
[the remark this answers is not in the document]
Comment (GuyHarwood) [reply admin/controllers/authentication.js]: this is now obsolete, i will remove

[the remark this answers is not in the document]
Comment (GuyHarwood) [reply admin/controllers/authentication.js]: this method is now obsolete, and no longer used.  will remove

File: admin/app.js (hunk 8 above)
Comment (GuyHarwood) [reply admin/app.js]: i've added a synchronous implementation, but it fails sporadically on discovery, likely due to the signin service configuring short request timeouts.

File: admin/controllers/authentication.js (hunk 9 above)
Comment (GuyHarwood) [reply admin/controllers/authentication.js]: the implementation in `master` from @GeorgeChatzigiannis recent PR always routes helpdesk to the school home page, which in turn redirects to the impersonation page, if necessary - https://github.com/DFEAGILEDEVOPS/MTC/blob/43b8288ceb7ba4c374a59140317669c60e4b8f4d/admin/controllers/school.js#L27

Seems like a valid solution to me

File: admin/services/dfe-signin.service.js (hunk 10 above)
Comment (GuyHarwood) [reply admin/services/dfe-signin.service.js]: good catch.  i will remove the return and superfluous re-assignment.

File: admin/app.js (hunk 8 above)
Comment (GuyHarwood) [reply admin/app.js]: switched back to async, with promise based call in app.js

File: admin/app.js [hunk 12]
```diff
@@ -210,19 +212,34 @@ passport.deserializeUser(function (user, done) {
...
-  )
-)
+if (config.Auth.mode === authModes.dfeSignIn) {
+  dfeSignInStrategy.initialiseAsync()
+    .then((strategy) => {
+      passport.use(authModes.dfeSignIn, strategy)
+    })
+    .catch((error) => {
+      logger.error(`unable to configure passport for dfeSignin:${error.message}`)
+      process.exit(1)
```
Comment (jon-shipley) [inline admin/app.js]: Let's test this under load later on and see how it works out.  My concern is that at high-loads we could see our dependency (the Sign-in service) causing our app to fail.

I would prefer our app to handle lack of availability on the Sign-in service by doing something reasonable like showing the user a message asking them to try again later.  Getting that into the start-up phase seems like a lot more work if it's even possible at all.

This is fine for a version 1 though.

File: admin/app.js [hunk 13]
```diff
@@ -210,19 +212,34 @@ passport.deserializeUser(function (user, done) {
...
+      passport.use(authModes.local,
+        new LocalStrategy({
+          passReqToCallback: true
+        },
+        require('./authentication/local-strategy')
+        )
+      )
+      break
+  }
+}
```
Comment (jon-shipley) [inline admin/app.js]: tidy 💯

File: admin/package.json [hunk 14]
```diff
@@ -22,10 +22,11 @@
     "parse-school-changes-csv": "node bin/parse-school-changes-csv.js"
   },
   "engines": {
-    "node": ">= 8.11"
+    "node": ">= 10.16.3"
```
Comment (jon-shipley) [inline admin/package.json]: Thanks for doing this

File: admin/services/data-access/dfe-signin.data.service.js [hunk 15]
```diff
@@ -0,0 +1,66 @@
...
+    logger.error(`dfe API call failed: statusCode:${response.status} - ${response.statusText}`)
+    throw new Error(`unsatisfactory response returned from DfE API. statusCode:${response.status}`)
+  }
+}
+
+/**
+ * @description creates a JWT signed with the api secret using the HS256 algorithm.
+ * @returns {Promise<string>} the signed Json Web Token
+ */
+const createJwtForDfeApi = async () => {
```
Comment (jon-shipley) [inline admin/services/data-access/dfe-signin.data.service.js]: Why is this method async?  It doesn't await.  The promise will still be returned from the `signAsync` method.

File: admin/spec/back-end/service/dfe-signin.service.spec.js [hunk 16]
```diff
@@ -0,0 +1,247 @@
...
+    try {
+      await sut.initialiseUser({ organisation: { urn: 12345 } }, token)
+      fail('expected error to be thrown')
+    } catch (error) {
+      expect(error).toBeDefined()
+      expect(error.message).toBe('unable to find user record')
+      done()
+    }
+  })
+  // failing WIP
```
Comment (jon-shipley) [inline admin/spec/back-end/service/dfe-signin.service.spec.js]: Is this comment still valid?

File: admin/services/data-access/sql.service.js [hunk 17]
```diff
@@ -37,7 +37,7 @@ let pool
  * @return {string | undefined}
  */
 const findDataType = (type) => Object.keys(sqlService.TYPES).find(k => {
-  logger.debug(`findDataType('${type}'): called`)
+  // logger.debug(`findDataType('${type}'): called`)
```
[the remark this answers is not in the document]
Comment (GuyHarwood) [reply admin/services/data-access/sql.service.js]: yes i would prefer to - as i mentioned on the session earlier this week, i want a new logger category for sql output, at the debug level.  These would come back into play at that point.

File: admin/services/data-access/dfe-signin.data.service.js (hunk 15 above)
Comment (GuyHarwood) [reply admin/services/data-access/dfe-signin.data.service.js]: fixed

**Pull request merged**